File: javascript/problems/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

const twoSum = function(nums, target) {
    let hash = {};
    for (let i = 0; i < nums.length; i++) {
        const j = target - nums[i];
        if (typeof hash[j] !== 'undefined') {
            return [hash[j], i];
        }
        hash[nums[i]] = i;
    }
    return [];
};

console.log(twoSum([2,7,11,15], 9));
console.log(twoSum([3,2,4], 6));

File: golang/exercise/two_sum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	hash := make(map[int]int)
	for i, num := range nums {
		sub := target - num
		if _, ok := hash[sub]; ok {
			return []int{hash[sub], i}
		}
		hash[nums[i]] = i
	}
	return nil
}

func main() {
	fmt.Println(twoSum([]int{2, 7, 11, 15}, 9))
	fmt.Println(twoSum([]int{3, 2, 4}, 6))
	fmt.Println(twoSum([]int{3, 2, 4}, 9))
}
